aneekee/nodejs-aws-shop-backend
# task-7

Frontend -- https://d1vdv7tmd1pu8.cloudfront.net/
Frontend PR -- https://github.com/slnchn/nodejs-aws-shop-react/pull/5
Cognito Auth Page -- https://slnchnrssdevpool.auth.eu-west-1.amazoncognito.com/oauth2/authorize?client_id=1hrrj9p0s56el2mcd7ca3us3i1&response_type=code&scope=aws.cognito.signin.user.admin+email+openid+phone+profile&redirect_uri=https%3A%2F%2Fd1vdv7tmd1pu8.cloudfront.net%2F

## How to sign-in with Cognito (optional)

1. Go https://slnchnrssdevpool.auth.eu-west-1.amazoncognito.com/oauth2/authorize?client_id=1hrrj9p0s56el2mcd7ca3us3i1&response_type=code&scope=aws.cognito.signin.user.admin+email+openid+phone+profile&redirect_uri=https%3A%2F%2Fd1vdv7tmd1pu8.cloudfront.net%2F (preferably in Incognito mode)
2. If you haven't created a user in my app please click 'Sign Up'
3. Fill in the fields (please use the real email so you can confirm it)
4. Confirm the email
5. It should redirect you to my frontend app main page
6. If the products don't show up please reload the page a couple of times (I didn't have much time so it's not ideal)


## How to test

1. Go https://d1vdv7tmd1pu8.cloudfront.net/admin/products (preferably in Incognito mode)
2. Try importing a csv file
3. It's expected that it'll not work + show 401 alert
4. Go to the local storage
5. Add this pair to the local storage: `authorization_token`-`c2xuY2huOlRFU1RfUEFTU1dPUkQ=` (this is a token with my login and test password -- `slnchn:TEST_PASSWORD`)
6. Try to import a csv file again
7. It should work, but if you're not signed in with the Cognito Auth page you won't see your imported items in the product list (I'll leave a short guide on how to sign in with Cognito below)
8. Slightly change the token in the local storage (or maybe generate another token with your username and password, or maybe my username but wrong password)
9. Try to import a csv file again
10. It should not work and show 403 alert

If you have any questions feel free to text me in discord :) 

## [Tasks](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/07_authorization/task.md#task-7-authorization)

### Task 7.1

- [x] 1. Create a new service called authorization-service at the same level as Product and Import services with its own AWS CDK Stack.
- [x] 2. Create a lambda function called basicAuthorizer under the Authorization Service.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-6469e4065b8b6fe20d193333e44fe42427ee58894b73ea88b965fbd825d324efR64
- [x] 3. This lambda should have at least one environment variable with the credentials described in the task
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-3ea34a30eac1a40be4d8ad033cf1165f79f932788464f830ae258bb9f91ad6c1R22
- [x] 4. This basicAuthorizer lambda should take Basic Authorization token, decode it and check that credentials provided by token exist in the lambda environment variable.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-6469e4065b8b6fe20d193333e44fe42427ee58894b73ea88b965fbd825d324efR78-R108
- [x] 5. This lambda should return 403 HTTP status if access is denied for this user (invalid authorization_token) and 401 HTTP status if Authorization header is not provided.
returning 401 -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-6469e4065b8b6fe20d193333e44fe42427ee58894b73ea88b965fbd825d324efR81
returning 403 -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-6469e4065b8b6fe20d193333e44fe42427ee58894b73ea88b965fbd825d324efR95 and https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-6469e4065b8b6fe20d193333e44fe42427ee58894b73ea88b965fbd825d324efR101

### Task 7.2

- [x] 1. Add Lambda authorization to the /import path of the Import Service API Gateway.
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-1bdd219556819b1facd80ab9feaa9255befd5d057318c3eec3636ada3d813209R114
- [x] 2. Use your basicAuthorizer lambda as the Lambda authorizer
Done -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-1bdd219556819b1facd80ab9feaa9255befd5d057318c3eec3636ada3d813209R51-R55

### Task 7.3

- [x] 1. Request from the client application to the /import path of the Import Service should have Basic Authorization header
Done -- https://github.com/slnchn/nodejs-aws-shop-react/pull/5/files#diff-24432b1f8209e7604bd55964d5a41321169ae306bbaa65064d4ed9a03fdd4d2fR43-R45
- [x] 2. Client should get authorization_token value from browser localStorage
Done -- https://github.com/slnchn/nodejs-aws-shop-react/pull/5/files#diff-24432b1f8209e7604bd55964d5a41321169ae306bbaa65064d4ed9a03fdd4d2fR35

### Task 7.4

- [x] 1. Commit all your work to separate branch (e.g. task-7 from the latest master) in your own repository.
- [x] 2. Create a pull request to the master branch.
- [x] 3. Submit link to the pull request to Crosscheck page in RS App.

### Additional tasks

- [x] Client application should display alerts for the responses in 401 and 403 HTTP statuses. This behavior should be added to the nodejs-aws-fe-main/src/index.tsx file.
Done -- https://github.com/slnchn/nodejs-aws-shop-react/pull/5/files#diff-0b5adbfe7b36e4ae2f479291e20152e33e940f7f265162d77f40f6bdb5da7405R12-R24
Also, I had to configure Gateway Responses -- https://github.com/slnchn/nodejs-aws-shop-backend/pull/5/files#diff-1bdd219556819b1facd80ab9feaa9255befd5d057318c3eec3636ada3d813209R66-R86

- [x] Optional Cognito task
I did the backend part with AWS console. 
The client side -- https://github.com/slnchn/nodejs-aws-shop-react/pull/5/commits/bf2ef3cb40ba5f1558323170e50424073e560aec
